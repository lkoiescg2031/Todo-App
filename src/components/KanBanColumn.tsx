import React, { MouseEvent, PureComponent } from "react";

import { withStyles } from "@material-ui/core/styles";

import Card from "@material-ui/core/Card";
import CardHeader from "@material-ui/core/CardHeader";

import grey from "@material-ui/core/colors/grey";

import CardCreateButton from "./CardCreateButton";
import ColumnMenuButton from "./ColumnMenuButton";
import ColumnTitle from "./ColumnTitle";

interface KanBanColumnProp {
	name: string;
	classes: {
		columns: string;
		works: string;
	};
}

interface KanBanColumnState {
	isTitleEdit: boolean;
}

export const Width = "260px";

class KanBanColumns extends PureComponent<KanBanColumnProp, KanBanColumnState> {
	constructor(props: KanBanColumnProp) {
		super(props);

		this.state = {
			isTitleEdit: false,
		};
		this.onColumnNameClicked = this.onColumnNameClicked.bind(this);
	}

	onColumnNameClicked(e: MouseEvent) {
		this.setState((prevState) => ({
			...prevState,
			isTitleEdit: !prevState.isTitleEdit,
		}));
	}

	render() {
		const { classes, children, name } = this.props;
		const childNode = (children as Array<any>).filter(
			(child) => typeof child !== "boolean"
		)[0];

		const appendCardText =
			Array.isArray(childNode) && childNode.length === 0
				? "Add a card"
				: "Add another card";

		return (
			<Card elevation={2} classes={{ root: classes.columns }}>
				<CardHeader
					title={<ColumnTitle value={name} initialMode="text" />}
					action={<ColumnMenuButton />}
				/>
				<div className={classes.works}>
					{children}
					<CardCreateButton text={appendCardText} />
				</div>
			</Card>
		);
	}
}

export default withStyles((theme) => ({
	columns: {
		minWidth: Width,
		height: "max-content",
		margin: `0px ${theme.spacing(1)}px`,
		padding: theme.spacing(1),
		background: grey[100],
	},
	works: {
		"& > div": {
			margin: `${theme.spacing(1)}px 0px`,
		},
	},
}))(KanBanColumns);
