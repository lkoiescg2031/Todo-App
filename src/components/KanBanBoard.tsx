import React from "react";

import { makeStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import AddIcon from "@material-ui/icons/Add";
import { Width } from "./KanBanColumn";

interface KanBanBoardProps {
	children?: React.ReactNode;
}

const useKanBanBoardStyles = makeStyles((theme) => ({
	boardWrapper: {
		width: "100%",
		height: "100%",
		overflow: "auto",
	},
	board: {
		width: "max-content",
		display: "flex",
		padding: `${theme.spacing(2)}px ${theme.spacing(1)}px`,
	},
	addButton: {
		minWidth: Width,
		height: "max-content",
		margin: `0px ${theme.spacing(1)}px`,
		textTransform: "none",
		backgroundColor: "rgba(255,255,255, 0.2)",
	},
}));

export default function KanBanBoard({ children }: KanBanBoardProps) {
	const classes = useKanBanBoardStyles();
	const appendText =
		typeof children === "undefined" ? "Add list" : "Add another list";
	return (
		<div className={classes.boardWrapper}>
			<div className={classes.board}>
				{children}
				<Button classes={{ root: classes.addButton }} startIcon={<AddIcon />}>
					{appendText}
				</Button>
			</div>
		</div>
	);
}
