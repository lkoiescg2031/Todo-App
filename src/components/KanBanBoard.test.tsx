import React from "react";
import { render } from "@testing-library/react";

import KanBanBoard from "./KanBanBoard";
import KanBanColumn from "./KanBanColumn";

describe("<KanBanBoard />", () => {
	it("snapshot test", () => {
		const utils = render(
			<KanBanBoard>
				<KanBanColumn name="Done" />
			</KanBanBoard>
		);
		expect(utils.container).toMatchSnapshot();
	});

	it("no child append Button name test", () => {
		const appendButtonText = "Add list";
		const utils = render(<KanBanBoard></KanBanBoard>);

		utils.getByText(appendButtonText);
	});

	it("has children append button name test", () => {
		const columnNames = ["TODO", "In progress", "Done"];
		const appendButtonText = "Add another list";

		const utils = render(
			<KanBanBoard>
				{columnNames.map((name, idx) => (
					<KanBanColumn key={`Column-${idx}`} name={name} />
				))}
			</KanBanBoard>
		);

		utils.getByText(appendButtonText);
	});

	it("children render test", () => {
		const columnNames = ["TODO", "In progress", "Done"];

		const utils = render(
			<KanBanBoard>
				{columnNames.map((name, idx) => (
					<KanBanColumn key={`Column-${idx}`} name={name} />
				))}
			</KanBanBoard>
		);

		// 자식들이 모두 렌더 되었는 지 확인
		columnNames.forEach((name) => {
			utils.getByText(name);
		});
	});
});
