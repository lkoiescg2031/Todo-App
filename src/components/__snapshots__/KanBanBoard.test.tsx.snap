// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KanBanBoard /> snapshot test 1`] = `
<div>
  <div
    class="makeStyles-boardWrapper-1"
  >
    <div
      class="makeStyles-board-2"
    >
      <div
        class="MuiPaper-root MuiCard-root KanBanColumns-columns-4 MuiPaper-elevation2 MuiPaper-rounded"
      >
        <div
          class="MuiCardHeader-root"
        >
          <div
            class="MuiCardHeader-content"
          >
            <span
              class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
            >
              <span
                class="EditableText-spanText-7"
              >
                Done
              </span>
            </span>
          </div>
          <div
            class="MuiCardHeader-action"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="KanBanColumns-works-5"
        />
        <div
          class="MuiCardActions-root MuiCardActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root KanBanColumns-addButton-6 MuiButton-text MuiButton-fullWidth"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
              </span>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Add a card
              </p>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M4 6H2v14c0 1.1.9 2 2 2h14v-2H4V6z"
                />
                <path
                  d="M20 2H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 10l-2.5-1.5L15 12V4h5v8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-addButton-3 MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          Add another list
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
